### DISCUSSION TITLE
Idiomatic UDP solutions

### DISCUSSION THREAD

**Discussion by JKurzer** [ntex-rs/ntex#352] [Q&A]
Hello hello!

I have the grave misfortune of needing an ultra-low latency UDP solution that also has a concept of session separate from the individual "connection." I know. UDP's not really for that. Yes, no, yes, I know, UDP's not really for this. It's okay, I'll sign a waiver. I need that extra millisecond. For reasons. Videogame reasons. 

I'm a big fan of Ntex generally, especially for slightly-stateful request handling, but I was wondering if there was a good idiomatic way to use it for handling small UDP packets with fixed msg buffers?   
  
To be exact about my use-case, I have a circumstance where I'm building out a set of UDP reflectors that bounce traffic from one player to the others. Similar to steam datagram relays, but with some sharp constraints in the name of speed, and also, I'd like to have access to it on non-pc platforms.

- Session membership and authentication is established with an mTLS or TLS2+ TCP connection that's used as a backhaul for time sync and continued session management. 
- A sym key is then provisioned for the session as a whole. This sym key is shared with all players in the session.
- No sensitive information is transmitted along these streams, only controller input, and players never see the IPs of other players.
- Secondary sym keys are provisioned as needed to segregate data access if required.  
- This means that the K,K,V mapping of IP+SymKey->Destination Relay is sufficient to attest to membership in a session, allowing the relays to be very very stupid, and thus highly process parallel.  
- This is important because they are geographically dispersed, and not just because I had someone grenade my server room once.  
- All of this is basically trivial in Ntex, so I'd like to use Ntex. Yes, I know it's a little insane, but we've tested it and it's shockingly fast.    

So the hard part, sessions, is easy in NTex and rust. But I'd also like to handle the actual UDP reflection in an idiomatic way though. I know that I could just use core libs or similar, but I was sort of hoping this might be a solved or semi-solved problem. I could technically probably use Ntex's websockets, but I'd like to avoid making that a client requirement. I skimmed the docs, but I obviously might have missed something since I'm still skilling up on Rust. What's the best\normal\least-bad approach for the UDP handling?

**Comment by fafhrd91** [comment]
hello. udp story is completely undeveloped, you should not use ntex for UDP.

**Comment by 0xle0ne** [comment]
@fafhrd91 Would you be open to add the udp function to connect and listen to ntex_rt ?

**Reply by fafhrd91** [reply]
ntex provides abstraction layer on top of stream sockets. there is no such functionality for udp, at the moment there is no reason to add anything related to ntex. but you can use udp directly from host runtime though.

**Reply by kylak** [reply]
What's about trying to use, on ntex, the "quiche" crate by Cloudfare in order to use HTTP 3 ?
Do you think it would be possible even if ntex has not implemented UDP so far ?

The salvo Rust web server framework uses the crate "quinn" in order to manage HTTP 3.

@fafhrd91 @fakeshadow
